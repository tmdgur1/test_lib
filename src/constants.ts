export const InjectionSymbols = {
  MessageRepository: Symbol('MessageRepository'),
  StorageService: Symbol('StorageService'),
  Logger: Symbol('Logger'),
  AgentContextProvider: Symbol('AgentContextProvider'),
  AgentDependencies: Symbol('AgentDependencies'),
  Stop$: Symbol('Stop$'),
  FileSystem: Symbol('FileSystem'),
  Wallet: Symbol('Wallet'),
}

export const DID_COMM_TRANSPORT_QUEUE = 'didcomm:transport/queue'
