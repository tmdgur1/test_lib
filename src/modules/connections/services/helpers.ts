import type { DidDocument } from '../../dids'
import type { DidDoc, PublicKey } from '../models'

import { Key, KeyType } from '../../../crypto'
import { AriesFrameworkError } from '../../../error'
import { IndyAgentService, DidCommV1Service, DidDocumentBuilder, getEd25519VerificationKey2018 } from '../../dids'
import { didDocumentJsonToNumAlgo1Did } from '../../dids/methods/peer/peerDidNumAlgo1'
import { EmbeddedAuthentication } from '../models'

export function convertToNewDidDocument(didDoc: DidDoc): DidDocument {
  const didDocumentBuilder = new DidDocumentBuilder('')

  const oldIdNewIdMapping: { [key: string]: string } = {}

  didDoc.authentication.forEach((auth) => {
    const { publicKey: pk } = auth

    // did:peer did documents can only use referenced keys.
    if (pk.type === 'Ed25519VerificationKey2018' && pk.value) {
      const ed25519VerificationMethod = convertPublicKeyToVerificationMethod(pk)

      const oldKeyId = normalizeId(pk.id)
      oldIdNewIdMapping[oldKeyId] = ed25519VerificationMethod.id
      didDocumentBuilder.addAuthentication(ed25519VerificationMethod.id)

      // Only the auth is embedded, we also need to add the key to the verificationMethod
      // for referenced authentication this should already be the case
      if (auth instanceof EmbeddedAuthentication) {
        didDocumentBuilder.addVerificationMethod(ed25519VerificationMethod)
      }
    }
  })

  didDoc.publicKey.forEach((pk) => {
    if (pk.type === 'Ed25519VerificationKey2018' && pk.value) {
      const ed25519VerificationMethod = convertPublicKeyToVerificationMethod(pk)

      const oldKeyId = normalizeId(pk.id)
      oldIdNewIdMapping[oldKeyId] = ed25519VerificationMethod.id
      didDocumentBuilder.addVerificationMethod(ed25519VerificationMethod)
    }
  })

  didDoc.didCommServices.forEach((service) => {
    const serviceId = normalizeId(service.id)

    // For didcommv1, we need to replace the old id with the new ones
    if (service instanceof DidCommV1Service) {
      const recipientKeys = service.recipientKeys.map((keyId) => {
        const oldKeyId = normalizeId(keyId)
        return oldIdNewIdMapping[oldKeyId]
      })

      service = new DidCommV1Service({
        id: serviceId,
        recipientKeys,
        serviceEndpoint: service.serviceEndpoint,
        routingKeys: service.routingKeys,
        accept: service.accept,
        priority: service.priority,
      })
    } else if (service instanceof IndyAgentService) {
      service = new IndyAgentService({
        id: serviceId,
        recipientKeys: service.recipientKeys,
        serviceEndpoint: service.serviceEndpoint,
        routingKeys: service.routingKeys,
        priority: service.priority,
      })
    }

    didDocumentBuilder.addService(service)
  })

  const didDocument = didDocumentBuilder.build()

  const peerDid = didDocumentJsonToNumAlgo1Did(didDocument.toJSON())
  didDocument.id = peerDid

  return didDocument
}

function normalizeId(fullId: string): `#${string}` {
  // Some old dids use `;` as the delimiter for the id. If we can't find a `#`
  // and a `;` exists, we will parse everything after `;` as the id.
  if (!fullId.includes('#') && fullId.includes(';')) {
    const [, ...ids] = fullId.split(';')

    return `#${ids.join(';')}`
  }

  const [, ...ids] = fullId.split('#')
  return `#${ids.length ? ids.join('#') : fullId}`
}

function convertPublicKeyToVerificationMethod(publicKey: PublicKey) {
  if (!publicKey.value) {
    throw new AriesFrameworkError(`Public key ${publicKey.id} does not have value property`)
  }
  const publicKeyBase58 = publicKey.value
  const ed25519Key = Key.fromPublicKeyBase58(publicKeyBase58, KeyType.Ed25519)
  return getEd25519VerificationKey2018({
    id: `#${publicKeyBase58.slice(0, 8)}`,
    key: ed25519Key,
    controller: '#id',
  })
}
