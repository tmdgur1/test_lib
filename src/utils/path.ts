/**
 * Extract directory from path (should also work with windows paths)
 *
 * @param path the path to extract the directory from
 * @returns the directory path
 */
export function getDirFromFilePath(path: string) {
  return path.substring(0, Math.max(path.lastIndexOf('/'), path.lastIndexOf('\\')))
}
