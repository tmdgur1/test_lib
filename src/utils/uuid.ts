import { v4, validate } from 'uuid'

export function uuid() {
  return v4()
}

export function isValidUuid(id: string) {
  return validate(id)
}
