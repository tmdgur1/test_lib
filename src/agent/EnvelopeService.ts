import type { AgentMessage } from './AgentMessage'
import type { AgentContext } from './context'
import type { EncryptedMessage, PlaintextMessage } from '../types'

import { InjectionSymbols } from '../constants'
import { Key, KeyType } from '../crypto'
import { Logger } from '../logger'
import { ForwardMessage } from '../modules/routing/messages'
import { inject, injectable } from '../plugins'

export interface EnvelopeKeys {
  recipientKeys: Key[]
  routingKeys: Key[]
  senderKey: Key | null
}

@injectable()
export class EnvelopeService {
  private logger: Logger

  public constructor(@inject(InjectionSymbols.Logger) logger: Logger) {
    this.logger = logger
  }

  public async packMessage(
    agentContext: AgentContext,
    payload: AgentMessage,
    keys: EnvelopeKeys
  ): Promise<EncryptedMessage> {
    const { recipientKeys, routingKeys, senderKey } = keys
    let recipientKeysBase58 = recipientKeys.map((key) => key.publicKeyBase58)
    const routingKeysBase58 = routingKeys.map((key) => key.publicKeyBase58)
    const senderKeyBase58 = senderKey && senderKey.publicKeyBase58

    // pass whether we want to use legacy did sov prefix
    const message = payload.toJSON({ useDidSovPrefixWhereAllowed: agentContext.config.useDidSovPrefixWhereAllowed })

    this.logger.debug(`Pack outbound message ${message['@type']}`)

    let encryptedMessage = await agentContext.wallet.pack(message, recipientKeysBase58, senderKeyBase58 ?? undefined)

    // If the message has routing keys (mediator) pack for each mediator
    for (const routingKeyBase58 of routingKeysBase58) {
      const forwardMessage = new ForwardMessage({
        // Forward to first recipient key
        to: recipientKeysBase58[0],
        message: encryptedMessage,
      })
      recipientKeysBase58 = [routingKeyBase58]
      this.logger.debug('Forward message created', forwardMessage)

      const forwardJson = forwardMessage.toJSON({
        useDidSovPrefixWhereAllowed: agentContext.config.useDidSovPrefixWhereAllowed,
      })

      // Forward messages are anon packed
      encryptedMessage = await agentContext.wallet.pack(forwardJson, [routingKeyBase58], undefined)
    }

    return encryptedMessage
  }

  public async unpackMessage(
    agentContext: AgentContext,
    encryptedMessage: EncryptedMessage
  ): Promise<DecryptedMessageContext> {
    const decryptedMessage = await agentContext.wallet.unpack(encryptedMessage)
    const { recipientKey, senderKey, plaintextMessage } = decryptedMessage
    return {
      recipientKey: recipientKey ? Key.fromPublicKeyBase58(recipientKey, KeyType.Ed25519) : undefined,
      senderKey: senderKey ? Key.fromPublicKeyBase58(senderKey, KeyType.Ed25519) : undefined,
      plaintextMessage,
    }
  }
}

export interface DecryptedMessageContext {
  plaintextMessage: PlaintextMessage
  senderKey?: Key
  recipientKey?: Key
}
