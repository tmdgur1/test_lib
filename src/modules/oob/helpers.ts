import type { OutOfBandInvitationOptions } from './messages'

import { ConnectionInvitationMessage, HandshakeProtocol } from '../connections'
import { didKeyToVerkey, verkeyToDidKey } from '../dids/helpers'

import { OutOfBandDidCommService } from './domain/OutOfBandDidCommService'
import { OutOfBandInvitation } from './messages'

export function convertToNewInvitation(oldInvitation: ConnectionInvitationMessage) {
  let service

  if (oldInvitation.did) {
    service = oldInvitation.did
  } else if (oldInvitation.serviceEndpoint && oldInvitation.recipientKeys && oldInvitation.recipientKeys.length > 0) {
    service = new OutOfBandDidCommService({
      id: '#inline',
      recipientKeys: oldInvitation.recipientKeys?.map(verkeyToDidKey),
      routingKeys: oldInvitation.routingKeys?.map(verkeyToDidKey),
      serviceEndpoint: oldInvitation.serviceEndpoint,
    })
  } else {
    throw new Error('Missing required serviceEndpoint, routingKeys and/or did fields in connection invitation')
  }

  const options: OutOfBandInvitationOptions = {
    id: oldInvitation.id,
    label: oldInvitation.label,
    imageUrl: oldInvitation.imageUrl,
    appendedAttachments: oldInvitation.appendedAttachments,
    accept: ['didcomm/aip1', 'didcomm/aip2;env=rfc19'],
    services: [service],
    handshakeProtocols: [HandshakeProtocol.Connections],
  }

  return new OutOfBandInvitation(options)
}

export function convertToOldInvitation(newInvitation: OutOfBandInvitation) {
  // Taking first service, as we can only include one service in a legacy invitation.
  const [service] = newInvitation.getServices()

  let options
  if (typeof service === 'string') {
    options = {
      id: newInvitation.id,
      label: newInvitation.label,
      did: service,
      imageUrl: newInvitation.imageUrl,
      appendedAttachments: newInvitation.appendedAttachments,
    }
  } else {
    options = {
      id: newInvitation.id,
      label: newInvitation.label,
      recipientKeys: service.recipientKeys.map(didKeyToVerkey),
      routingKeys: service.routingKeys?.map(didKeyToVerkey),
      serviceEndpoint: service.serviceEndpoint,
      imageUrl: newInvitation.imageUrl,
      appendedAttachments: newInvitation.appendedAttachments,
    }
  }

  const connectionInvitationMessage = new ConnectionInvitationMessage(options)
  return connectionInvitationMessage
}
