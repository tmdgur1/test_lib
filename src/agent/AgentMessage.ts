import type { ParsedMessageType } from '../utils/messageType'
import type { Constructor } from '../utils/mixins'

import { Exclude } from 'class-transformer'

import { AckDecorated } from '../decorators/ack/AckDecoratorExtension'
import { AttachmentDecorated } from '../decorators/attachment/AttachmentExtension'
import { L10nDecorated } from '../decorators/l10n/L10nDecoratorExtension'
import { ServiceDecorated } from '../decorators/service/ServiceDecoratorExtension'
import { ThreadDecorated } from '../decorators/thread/ThreadDecoratorExtension'
import { TimingDecorated } from '../decorators/timing/TimingDecoratorExtension'
import { TransportDecorated } from '../decorators/transport/TransportDecoratorExtension'
import { JsonTransformer } from '../utils/JsonTransformer'
import { replaceNewDidCommPrefixWithLegacyDidSovOnMessage } from '../utils/messageType'

import { BaseMessage } from './BaseMessage'

export type ConstructableAgentMessage = Constructor<AgentMessage> & { type: ParsedMessageType }

const Decorated = ThreadDecorated(
  L10nDecorated(TransportDecorated(TimingDecorated(AckDecorated(AttachmentDecorated(ServiceDecorated(BaseMessage))))))
)

export class AgentMessage extends Decorated {
  /**
   * Whether the protocol RFC was initially written using the legacy did:prefix instead of the
   * new https://didcomm.org message type prefix.
   *
   * @see https://github.com/hyperledger/aries-rfcs/blob/main/features/0348-transition-msg-type-to-https/README.md
   */
  @Exclude()
  public readonly allowDidSovPrefix: boolean = false

  public toJSON({ useDidSovPrefixWhereAllowed }: { useDidSovPrefixWhereAllowed?: boolean } = {}): Record<
    string,
    unknown
  > {
    const json = JsonTransformer.toJSON(this)

    // If we have `useDidSovPrefixWhereAllowed` enabled, we want to replace the new https://didcomm.org prefix with the legacy did:sov prefix.
    // However, we only do this if the protocol RFC was initially written with the did:sov message type prefix
    // See https://github.com/hyperledger/aries-rfcs/blob/main/features/0348-transition-msg-type-to-https/README.md
    if (this.allowDidSovPrefix && useDidSovPrefixWhereAllowed) {
      replaceNewDidCommPrefixWithLegacyDidSovOnMessage(json)
    }

    return json
  }

  public is<C extends typeof AgentMessage>(Class: C): this is InstanceType<C> {
    return this.type === Class.type.messageTypeUri
  }
}
