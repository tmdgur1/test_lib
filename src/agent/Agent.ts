import type { AgentDependencies } from './AgentDependencies'
import type { AgentModulesInput } from './AgentModules'
import type { AgentMessageReceivedEvent } from './Events'
import type { Module } from '../plugins'
import type { InboundTransport } from '../transport/InboundTransport'
import type { OutboundTransport } from '../transport/OutboundTransport'
import type { InitConfig } from '../types'
import type { Subscription } from 'rxjs'

import { Subject } from 'rxjs'
import { concatMap, takeUntil } from 'rxjs/operators'

import { InjectionSymbols } from '../constants'
import { SigningProviderToken } from '../crypto'
import { JwsService } from '../crypto/JwsService'
import { AriesFrameworkError } from '../error'
import { DependencyManager } from '../plugins'
import { DidCommMessageRepository, StorageUpdateService, StorageVersionRepository } from '../storage'
import { InMemoryMessageRepository } from '../storage/InMemoryMessageRepository'

import { AgentConfig } from './AgentConfig'
import { extendModulesWithDefaultModules } from './AgentModules'
import { BaseAgent } from './BaseAgent'
import { Dispatcher } from './Dispatcher'
import { EnvelopeService } from './EnvelopeService'
import { EventEmitter } from './EventEmitter'
import { AgentEventTypes } from './Events'
import { FeatureRegistry } from './FeatureRegistry'
import { MessageHandlerRegistry } from './MessageHandlerRegistry'
import { MessageReceiver } from './MessageReceiver'
import { MessageSender } from './MessageSender'
import { TransportService } from './TransportService'
import { AgentContext, DefaultAgentContextProvider } from './context'

interface AgentOptions<AgentModules extends AgentModulesInput> {
  config: InitConfig
  modules?: AgentModules
  dependencies: AgentDependencies
}

// Any makes sure you can use Agent as a type without always needing to specify the exact generics for the agent
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class Agent<AgentModules extends AgentModulesInput = any> extends BaseAgent<AgentModules> {
  public messageSubscription: Subscription

  public constructor(options: AgentOptions<AgentModules>, dependencyManager = new DependencyManager()) {
    console.log('[Agent constructor]', 1)
    const agentConfig = new AgentConfig(options.config, options.dependencies)
    console.log('[Agent constructor]', 2)
    const modulesWithDefaultModules = extendModulesWithDefaultModules(options.modules)
    
    // Register internal dependencies
    console.log('[Agent constructor]', 3)
    dependencyManager.registerSingleton(MessageHandlerRegistry)
    console.log('[Agent constructor]', 4)
    dependencyManager.registerSingleton(EventEmitter)
    console.log('[Agent constructor]', 5)
    dependencyManager.registerSingleton(MessageSender)
    console.log('[Agent constructor]', 6)
    dependencyManager.registerSingleton(MessageReceiver)
    console.log('[Agent constructor]', 7)
    dependencyManager.registerSingleton(TransportService)
    console.log('[Agent constructor]', 8)
    dependencyManager.registerSingleton(Dispatcher)
    console.log('[Agent constructor]', 9)
    dependencyManager.registerSingleton(EnvelopeService)
    console.log('[Agent constructor]', 10)
    dependencyManager.registerSingleton(FeatureRegistry)
    console.log('[Agent constructor]', 11)
    dependencyManager.registerSingleton(JwsService)
    console.log('[Agent constructor]', 12)
    dependencyManager.registerSingleton(DidCommMessageRepository)
    console.log('[Agent constructor]', 13)
    dependencyManager.registerSingleton(StorageVersionRepository)
    console.log('[Agent constructor]', 14)
    dependencyManager.registerSingleton(StorageUpdateService)
    
    // This is a really ugly hack to make tsyringe work without any SigningProviders registered
    // It is currently impossible to use @injectAll if there are no instances registered for the
    // token. We register a value of `default` by default and will filter that out in the registry.
    // Once we have a signing provider that should always be registered we can remove this. We can make an ed25519
    // signer using the @stablelib/ed25519 library.
    console.log('[Agent constructor]', 15)
    dependencyManager.registerInstance(SigningProviderToken, 'default')
    
    console.log('[Agent constructor]', 16)
    dependencyManager.registerInstance(AgentConfig, agentConfig)
    console.log('[Agent constructor]', 17)
    dependencyManager.registerInstance(InjectionSymbols.AgentDependencies, agentConfig.agentDependencies)
    console.log('[Agent constructor]', 18)
    dependencyManager.registerInstance(InjectionSymbols.Stop$, new Subject<boolean>())
    console.log('[Agent constructor]', 19)
    dependencyManager.registerInstance(InjectionSymbols.FileSystem, new agentConfig.agentDependencies.FileSystem())
    
    // Register all modules. This will also include the default modules
    console.log('[Agent constructor]', 20)
    dependencyManager.registerModules(modulesWithDefaultModules)
    console.log('[Agent constructor]', 21)
    
    // Register possibly already defined services
    if (!dependencyManager.isRegistered(InjectionSymbols.Wallet)) {
      throw new AriesFrameworkError(
        "Missing required dependency: 'Wallet'. You can register it using one of the provided modules such as the AskarModule or the IndySdkModule, or implement your own."
        )
      }
    console.log('[Agent constructor]', 22)
    if (!dependencyManager.isRegistered(InjectionSymbols.Logger)) {
      dependencyManager.registerInstance(InjectionSymbols.Logger, agentConfig.logger)
    }
    console.log('[Agent constructor]', 23)
    if (!dependencyManager.isRegistered(InjectionSymbols.StorageService)) {
      throw new AriesFrameworkError(
        "Missing required dependency: 'StorageService'. You can register it using one of the provided modules such as the AskarModule or the IndySdkModule, or implement your own."
        )
      }
    console.log('[Agent constructor]', 24)
    if (!dependencyManager.isRegistered(InjectionSymbols.MessageRepository)) {
      dependencyManager.registerSingleton(InjectionSymbols.MessageRepository, InMemoryMessageRepository)
    }
    
    // TODO: contextCorrelationId for base wallet
    // Bind the default agent context to the container for use in modules etc.
    console.log('[Agent constructor]', 25)
    dependencyManager.registerInstance(
      AgentContext,
      new AgentContext({
        dependencyManager,
        contextCorrelationId: 'default',
      })
    )
      
      // If no agent context provider has been registered we use the default agent context provider.
    console.log('[Agent constructor]', 26)
    if (!dependencyManager.isRegistered(InjectionSymbols.AgentContextProvider)) {
      dependencyManager.registerSingleton(InjectionSymbols.AgentContextProvider, DefaultAgentContextProvider)
    }
    
    console.log('[Agent constructor]', 27)
    super(agentConfig, dependencyManager)
    
    console.log('[Agent constructor]', 28)
    const stop$ = this.dependencyManager.resolve<Subject<boolean>>(InjectionSymbols.Stop$)
    
    // Listen for new messages (either from transports or somewhere else in the framework / extensions)
    console.log('[Agent constructor]', 29)
    this.messageSubscription = this.eventEmitter
    .observable<AgentMessageReceivedEvent>(AgentEventTypes.AgentMessageReceived)
    .pipe(
      takeUntil(stop$),
      concatMap((e) =>
      this.messageReceiver
      .receiveMessage(e.payload.message, {
        connection: e.payload.connection,
        contextCorrelationId: e.payload.contextCorrelationId,
      })
      .catch((error) => {
        this.logger.error('Failed to process message', { error })
      })
      )
      )
      .subscribe()
    console.log('[Agent constructor]', 30)
  }

  public registerInboundTransport(inboundTransport: InboundTransport) {
    this.messageReceiver.registerInboundTransport(inboundTransport)
  }

  public async unregisterInboundTransport(inboundTransport: InboundTransport) {
    await this.messageReceiver.unregisterInboundTransport(inboundTransport)
  }

  public get inboundTransports() {
    return this.messageReceiver.inboundTransports
  }

  public registerOutboundTransport(outboundTransport: OutboundTransport) {
    this.messageSender.registerOutboundTransport(outboundTransport)
  }

  public async unregisterOutboundTransport(outboundTransport: OutboundTransport) {
    await this.messageSender.unregisterOutboundTransport(outboundTransport)
  }

  public get outboundTransports() {
    return this.messageSender.outboundTransports
  }

  public get events() {
    return this.eventEmitter
  }

  /**
   * Agent's feature registry
   */
  public get features() {
    return this.featureRegistry
  }

  public async initialize() {
    console.log('[Agent]', 1)
    await super.initialize()
    console.log('[Agent]', 2)
    
    for (const [, module] of Object.entries(this.dependencyManager.registeredModules) as [string, Module][]) {
      if (module.initialize) {
        await module.initialize(this.agentContext)
      }
    }
    console.log('[Agent]', 3)
    
    for (const transport of this.inboundTransports) {
      await transport.start(this)
    }
    console.log('[Agent]', 4)
    
    for (const transport of this.outboundTransports) {
      await transport.start(this)
    }
    console.log('[Agent]', 5)
    
    // Connect to mediator through provided invitation if provided in config
    // Also requests mediation ans sets as default mediator
    // Because this requires the connections module, we do this in the agent constructor
    if (this.mediationRecipient.config.mediatorInvitationUrl) {
      this.logger.debug('Provision mediation with invitation', {
        mediatorInvitationUrl: this.mediationRecipient.config.mediatorInvitationUrl,
      })
      const mediationConnection = await this.getMediationConnection(
        this.mediationRecipient.config.mediatorInvitationUrl
        )
        await this.mediationRecipient.provision(mediationConnection)
      }
      console.log('[Agent]', 6)
      await this.mediator.initialize()
      console.log('[Agent]', 7)
      await this.mediationRecipient.initialize()
      console.log('[Agent]', 8)
      
      this._isInitialized = true
      console.log('[Agent]', 9)
    }
    
    
  public async shutdown() {
    const stop$ = this.dependencyManager.resolve<Subject<boolean>>(InjectionSymbols.Stop$)
    // All observables use takeUntil with the stop$ observable
    // this means all observables will stop running if a value is emitted on this observable
    stop$.next(true)

    // Stop transports
    const allTransports = [...this.inboundTransports, ...this.outboundTransports]
    const transportPromises = allTransports.map((transport) => transport.stop())
    await Promise.all(transportPromises)

    if (this.wallet.isInitialized) {
      await this.wallet.close()
    }

    this._isInitialized = false
  }

  protected async getMediationConnection(mediatorInvitationUrl: string) {
    const outOfBandInvitation = await this.oob.parseInvitation(mediatorInvitationUrl)
    const outOfBandRecord = await this.oob.findByReceivedInvitationId(outOfBandInvitation.id)
    const [connection] = outOfBandRecord ? await this.connections.findAllByOutOfBandId(outOfBandRecord.id) : []

    if (!connection) {
      this.logger.debug('Mediation connection does not exist, creating connection')
      // We don't want to use the current default mediator when connecting to another mediator
      const routing = await this.mediationRecipient.getRouting({ useDefaultMediator: false })

      this.logger.debug('Routing created', routing)
      const { connectionRecord: newConnection } = await this.oob.receiveInvitation(outOfBandInvitation, {
        routing,
      })
      this.logger.debug(`Mediation invitation processed`, { outOfBandInvitation })

      if (!newConnection) {
        throw new AriesFrameworkError('No connection record to provision mediation.')
      }

      return this.connections.returnWhenIsConnected(newConnection.id)
    }

    if (!connection.isReady) {
      return this.connections.returnWhenIsConnected(connection.id)
    }
    return connection
  }
}
