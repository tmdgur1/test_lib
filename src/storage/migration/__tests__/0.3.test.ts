import { readFileSync } from 'fs'
import path from 'path'

import { InMemoryStorageService } from '../../../../../../tests/InMemoryStorageService'
import { IndySdkWallet } from '../../../../../indy-sdk/src'
import { IndySdkSymbol } from '../../../../../indy-sdk/src/types'
import { indySdk } from '../../../../tests'
import { agentDependencies } from '../../../../tests/helpers'
import { Agent } from '../../../agent/Agent'
import { InjectionSymbols } from '../../../constants'
import { DependencyManager } from '../../../plugins'
import * as uuid from '../../../utils/uuid'
import { UpdateAssistant } from '../UpdateAssistant'

const backupDate = new Date('2023-03-18T22:50:20.522Z')
jest.useFakeTimers().setSystemTime(backupDate)

const walletConfig = {
  id: `Wallet: 0.4 Update`,
  key: `Key: 0.4 Update`,
}

describe('UpdateAssistant | v0.3.1 - v0.4', () => {
  it(`should correctly update the did records and remove cache records`, async () => {
    // We need to mock the uuid generation to make sure we generate consistent uuids for the new records created.
    let uuidCounter = 1
    const uuidSpy = jest.spyOn(uuid, 'uuid').mockImplementation(() => `${uuidCounter++}-4e4f-41d9-94c4-f49351b811f1`)

    const aliceDidRecordsString = readFileSync(
      path.join(__dirname, '__fixtures__/alice-2-sov-dids-one-cache-record-0.3.json'),
      'utf8'
    )

    const dependencyManager = new DependencyManager()
    const storageService = new InMemoryStorageService()
    dependencyManager.registerInstance(InjectionSymbols.StorageService, storageService)
    // If we register the IndySdkModule it will register the storage service, but we use in memory storage here
    dependencyManager.registerContextScoped(InjectionSymbols.Wallet, IndySdkWallet)
    dependencyManager.registerInstance(IndySdkSymbol, indySdk)

    const agent = new Agent(
      {
        config: {
          label: 'Test Agent',
          walletConfig,
        },
        dependencies: agentDependencies,
      },
      dependencyManager
    )

    const updateAssistant = new UpdateAssistant(agent, {
      v0_1ToV0_2: {
        mediationRoleUpdateStrategy: 'doNotChange',
      },
    })

    await updateAssistant.initialize()

    // Set storage after initialization. This mimics as if this wallet
    // is opened as an existing wallet instead of a new wallet
    storageService.records = JSON.parse(aliceDidRecordsString)

    expect(await updateAssistant.isUpToDate()).toBe(false)
    expect(await updateAssistant.getNeededUpdates('0.4')).toEqual([
      {
        fromVersion: '0.3.1',
        toVersion: '0.4',
        doUpdate: expect.any(Function),
      },
    ])

    await updateAssistant.update()

    expect(await updateAssistant.isUpToDate()).toBe(true)
    expect(await updateAssistant.getNeededUpdates()).toEqual([])

    // MEDIATOR_ROUTING_RECORD recipientKeys will be different every time, and is not what we're testing here
    delete storageService.records.MEDIATOR_ROUTING_RECORD
    expect(storageService.records).toMatchSnapshot()

    await agent.shutdown()
    await agent.wallet.delete()

    uuidSpy.mockReset()
  })
})
