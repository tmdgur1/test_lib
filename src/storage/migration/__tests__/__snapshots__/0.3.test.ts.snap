// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAssistant | v0.3.1 - v0.4 should correctly update the did records and remove cache records 1`] = `
{
  "4993c740-5cd9-4c79-a7d8-23d1266d31be": {
    "id": "4993c740-5cd9-4c79-a7d8-23d1266d31be",
    "tags": {
      "did": "did:indy:bcovrin:test:Pow4pdnPgTS7JAXvWkoF2c",
      "legacyUnqualifiedDid": undefined,
      "method": "indy",
      "methodSpecificIdentifier": "bcovrin:test:Pow4pdnPgTS7JAXvWkoF2c",
      "qualifiedIndyDid": undefined,
      "recipientKeyFingerprints": [],
      "role": "created",
    },
    "type": "DidRecord",
    "value": {
      "createdAt": "2023-03-18T18:35:07.208Z",
      "did": "did:indy:bcovrin:test:Pow4pdnPgTS7JAXvWkoF2c",
      "id": "4993c740-5cd9-4c79-a7d8-23d1266d31be",
      "metadata": {},
      "role": "created",
      "updatedAt": "2023-03-18T22:50:20.522Z",
    },
  },
  "8168612b-73d1-4917-9a61-84e8102988f0": {
    "id": "8168612b-73d1-4917-9a61-84e8102988f0",
    "tags": {
      "did": "did:indy:bcovrin:test:8DFqUo6UtQLLZETE7Gm29k",
      "legacyUnqualifiedDid": undefined,
      "method": "indy",
      "methodSpecificIdentifier": "bcovrin:test:8DFqUo6UtQLLZETE7Gm29k",
      "qualifiedIndyDid": undefined,
      "recipientKeyFingerprints": [],
      "role": "created",
    },
    "type": "DidRecord",
    "value": {
      "createdAt": "2023-03-18T18:35:04.191Z",
      "did": "did:indy:bcovrin:test:8DFqUo6UtQLLZETE7Gm29k",
      "id": "8168612b-73d1-4917-9a61-84e8102988f0",
      "metadata": {},
      "role": "created",
      "updatedAt": "2023-03-18T22:50:20.522Z",
    },
  },
  "STORAGE_VERSION_RECORD_ID": {
    "id": "STORAGE_VERSION_RECORD_ID",
    "tags": {},
    "type": "StorageVersionRecord",
    "value": {
      "createdAt": "2023-03-18T18:35:02.888Z",
      "id": "STORAGE_VERSION_RECORD_ID",
      "metadata": {},
      "storageVersion": "0.4",
      "updatedAt": "2023-03-18T22:50:20.522Z",
    },
  },
}
`;
